// @electron-forge/maker-flatpak/dist/MakerFlatpak.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakerFlatpak = exports.flatpakArch = void 0;
const node_path_1 = __importDefault(require("node:path"));
const maker_base_1 = require("@electron-forge/maker-base");
const fs_extra_1 = __importDefault(require("fs-extra"));
function flatpakArch(nodeArch) {
    switch (nodeArch) {
        case 'ia32':
            return 'i386';
        case 'x64':
            return 'x86_64';
        case 'armv7l':
            return 'arm';
        case 'arm64':
            return 'aarch64';
        // arm => arm
        default:
            return nodeArch;
    }
}
exports.flatpakArch = flatpakArch;
class MakerFlatpak extends maker_base_1.MakerBase {
    constructor() {
        super(...arguments);
        this.name = 'flatpak';
        this.defaultPlatforms = ['linux'];
        this.requiredExternalBinaries = ['flatpak-builder', 'eu-strip'];
    }
    isSupportedOnCurrentPlatform() {
        return this.isInstalled('@malept/electron-installer-flatpak');
    }
    async make({ dir, makeDir, targetArch }) {
        // eslint-disable-next-line n/no-missing-require
        const installer = require('@malept/electron-installer-flatpak');
        const arch = flatpakArch(targetArch);
        const outDir = node_path_1.default.resolve(makeDir, 'flatpak', arch);
        await this.ensureDirectory(outDir);
        const flatpakConfig = {
            ...this.config,
            arch,
            src: dir,
            dest: outDir,
        };
        await installer(flatpakConfig);
        return (await fs_extra_1.default.readdir(outDir))
            .filter((basename) => basename.endsWith('.flatpak'))
            .map((basename) => node_path_1.default.join(outDir, basename));
    }
}
exports.default = MakerFlatpak;
exports.MakerFlatpak = MakerFlatpak;
