// @eyhn/msgpack-stream/dist.es2015+esm/index.mjs
// Main Functions:
import { encode, encodeStream } from "./encode.mjs";
export { encode, encodeStream };
import { decode, decodeMulti, decodeStream } from "./decode.mjs";
export { decode, decodeMulti, decodeStream };
import { decodeAsync, decodeArrayStream, decodeMultiStream } from "./decodeAsync.mjs";
export { decodeAsync, decodeArrayStream, decodeMultiStream };
import { Decoder, DataViewIndexOutOfBoundsError } from "./Decoder.mjs";
import { DecodeError } from "./DecodeError.mjs";
export { Decoder, DecodeError, DataViewIndexOutOfBoundsError };
import { Encoder } from "./Encoder.mjs";
export { Encoder };
// Utilitiies for Extension Types:
import { ExtensionCodec } from "./ExtensionCodec.mjs";
export { ExtensionCodec };
import { ExtData } from "./ExtData.mjs";
export { ExtData };
import { EXT_TIMESTAMP, encodeDateToTimeSpec, encodeTimeSpecToTimestamp, decodeTimestampToTimeSpec, encodeTimestampExtension, decodeTimestampExtension, } from "./timestamp.mjs";
export { EXT_TIMESTAMP, encodeDateToTimeSpec, encodeTimeSpecToTimestamp, decodeTimestampToTimeSpec, encodeTimestampExtension, decodeTimestampExtension, };
