// changelogithub/dist/index.cjs
'use strict';

const generate = require('./shared/changelogithub.CSEqPOk-.cjs');
require('changelogen');
require('semver');
require('ansis');
require('ofetch');
require('convert-gitmoji');



exports.defineConfig = generate.defineConfig;
exports.generate = generate.generate;
exports.generateMarkdown = generate.generateMarkdown;
exports.getCurrentGitBranch = generate.getCurrentGitBranch;
exports.getFirstGitCommit = generate.getFirstGitCommit;
exports.getGitHubRepo = generate.getGitHubRepo;
exports.getGitTags = generate.getGitTags;
exports.getLastMatchingTag = generate.getLastMatchingTag;
exports.getSafeTagTemplate = generate.getSafeTagTemplate;
exports.hasTagOnGitHub = generate.hasTagOnGitHub;
exports.isPrerelease = generate.isPrerelease;
exports.isRefGitTag = generate.isRefGitTag;
exports.isRepoShallow = generate.isRepoShallow;
exports.parseCommits = generate.parseCommits;
exports.resolveAuthorInfo = generate.resolveAuthorInfo;
exports.resolveAuthors = generate.resolveAuthors;
exports.resolveConfig = generate.resolveConfig;
exports.sendRelease = generate.sendRelease;
