// @toeverything/pdf-viewer-types/dist/index.js
/**
 * PDFium Error Code.
 */
export var ErrorCode;
(function (ErrorCode) {
    // No error.
    ErrorCode[ErrorCode["SUCCESS"] = 0] = "SUCCESS";
    // Unknown error.
    ErrorCode[ErrorCode["UNKNOWN"] = 1] = "UNKNOWN";
    // File not found or could not be opened.
    ErrorCode[ErrorCode["FILE"] = 2] = "FILE";
    // File not in PDF format or corrupted.
    ErrorCode[ErrorCode["FORMAT"] = 3] = "FORMAT";
    // Password required or incorrect password.
    ErrorCode[ErrorCode["PASSWORD"] = 4] = "PASSWORD";
    // Unsupported security scheme.
    ErrorCode[ErrorCode["SECURITY"] = 5] = "SECURITY";
    // Page not found or content error.
    ErrorCode[ErrorCode["PAGE"] = 6] = "PAGE";
})(ErrorCode || (ErrorCode = {}));
/**
 * PDFium Bitmap Format.
 * More DIB formats.
 */
export var BitmapFormat;
(function (BitmapFormat) {
    // Unknown or unsupported format.
    BitmapFormat[BitmapFormat["Unknown"] = 0] = "Unknown";
    // Gray scale bitmap, one byte per pixel.
    BitmapFormat[BitmapFormat["Gray"] = 1] = "Gray";
    // 3 bytes per pixel, byte order: blue, green, red.
    BitmapFormat[BitmapFormat["BGR"] = 2] = "BGR";
    // 4 bytes per pixel, byte order: blue, green, red, unused.
    BitmapFormat[BitmapFormat["BGRx"] = 3] = "BGRx";
    // 4 bytes per pixel, byte order: blue, green, red, alpha.
    BitmapFormat[BitmapFormat["BGRA"] = 4] = "BGRA";
})(BitmapFormat || (BitmapFormat = {}));
/**
 * Metadata Tag.
 */
export var MetaTag;
(function (MetaTag) {
    MetaTag["Title"] = "Title";
    MetaTag["Author"] = "Author";
    MetaTag["Subject"] = "Subject";
    MetaTag["Keywords"] = "Keywords";
    MetaTag["Creator"] = "Creator";
    MetaTag["Producer"] = "Producer";
    MetaTag["CreationDate"] = "CreationDate";
    MetaTag["ModificationDate"] = "ModificationDate";
})(MetaTag || (MetaTag = {}));
export const MetaTags = Object.values(MetaTag);
/**
 * Page rendering flags. They can be combined with bit-wise OR.
 */
export var PageRenderingflags;
(function (PageRenderingflags) {
    // Set if annotations are to be rendered.
    PageRenderingflags[PageRenderingflags["ANNOT"] = 1] = "ANNOT";
    // Set if using text rendering optimized for LCD display. This flag will only
    // take effect if anti-aliasing is enabled for text.
    PageRenderingflags[PageRenderingflags["LCD_TEXT"] = 2] = "LCD_TEXT";
    // Don't use the native text output available on some platforms
    PageRenderingflags[PageRenderingflags["NO_NATIVETEXT"] = 4] = "NO_NATIVETEXT";
    // Grayscale output.
    PageRenderingflags[PageRenderingflags["GRAYSCALE"] = 8] = "GRAYSCALE";
    // Obsolete, has no effect, retained for compatibility.
    PageRenderingflags[PageRenderingflags["DEBUG_INFO"] = 128] = "DEBUG_INFO";
    // Obsolete, has no effect, retained for compatibility.
    PageRenderingflags[PageRenderingflags["NO_CATCH"] = 256] = "NO_CATCH";
    // Limit image cache size.
    PageRenderingflags[PageRenderingflags["RENDER_LIMITEDIMAGECACHE"] = 512] = "RENDER_LIMITEDIMAGECACHE";
    // Always use halftone for image stretching.
    PageRenderingflags[PageRenderingflags["RENDER_FORCEHALFTONE"] = 1024] = "RENDER_FORCEHALFTONE";
    // Render for printing.
    PageRenderingflags[PageRenderingflags["PRINTING"] = 2048] = "PRINTING";
    // Set to disable anti-aliasing on text. This flag will also disable LCD
    // optimization for text rendering.
    PageRenderingflags[PageRenderingflags["RENDER_NO_SMOOTHTEXT"] = 4096] = "RENDER_NO_SMOOTHTEXT";
    // Set to disable anti-aliasing on images.
    PageRenderingflags[PageRenderingflags["RENDER_NO_SMOOTHIMAGE"] = 8192] = "RENDER_NO_SMOOTHIMAGE";
    // Set to disable anti-aliasing on paths.
    PageRenderingflags[PageRenderingflags["RENDER_NO_SMOOTHPATH"] = 16384] = "RENDER_NO_SMOOTHPATH";
    // Set whether to render in a reverse Byte order, this flag is only used when
    // rendering to a bitmap.
    PageRenderingflags[PageRenderingflags["REVERSE_BYTE_ORDER"] = 16] = "REVERSE_BYTE_ORDER";
    // Set whether fill paths need to be stroked. This flag is only used when
    // FPDF_COLORSCHEME is passed in, since with a single fill color for paths the
    // boundaries of adjacent fill paths are less visible.
    PageRenderingflags[PageRenderingflags["CONVERT_FILL_TO_STROKE"] = 32] = "CONVERT_FILL_TO_STROKE";
})(PageRenderingflags || (PageRenderingflags = {}));
export var FileIdentifier;
(function (FileIdentifier) {
    FileIdentifier[FileIdentifier["Permanent"] = 0] = "Permanent";
    FileIdentifier[FileIdentifier["Changing"] = 1] = "Changing";
})(FileIdentifier || (FileIdentifier = {}));
export var PageMode;
(function (PageMode) {
    // Unknown page mode.
    PageMode[PageMode["UNKNOWN"] = -1] = "UNKNOWN";
    // Document outline, and thumbnails hidden.
    PageMode[PageMode["USENONE"] = 0] = "USENONE";
    // Document outline visible.
    PageMode[PageMode["USEOUTLINES"] = 1] = "USEOUTLINES";
    // Thumbnail images visible.
    PageMode[PageMode["USETHUMBS"] = 2] = "USETHUMBS";
    // Full-screen mode, no menu bar, window controls, or other decorations visible.
    PageMode[PageMode["FULLSCREEN"] = 3] = "FULLSCREEN";
    // Optional content group panel visible.
    PageMode[PageMode["USEOC"] = 4] = "USEOC";
    // Attachments panel visible.
    PageMode[PageMode["USEATTACHMENTS"] = 5] = "USEATTACHMENTS";
})(PageMode || (PageMode = {}));
export const BYTES_PER_PIXEL = 4;
export var Rotation;
(function (Rotation) {
    // No rotation
    Rotation[Rotation["Zero"] = 0] = "Zero";
    // 90 degrees
    Rotation[Rotation["Quarter"] = 1] = "Quarter";
    // 180 degrees
    Rotation[Rotation["Half"] = 2] = "Half";
    // 270 degrees
    Rotation[Rotation["ThreeQuarters"] = 3] = "ThreeQuarters";
})(Rotation || (Rotation = {}));
