// @magic-works/i18n-codegen/dist/type.js
import { readFileSync } from 'node:fs';
import { dirname, join, relative } from 'node:path';
import ts from 'typescript';
export { GeneratorList, ParserList } from './json-schema.js';
export class ParserInput {
    constructor(sourceFile, originalFile, parserOptions) {
        this.sourceFile = sourceFile;
        this.originalFile = originalFile;
        this.parserOptions = parserOptions;
        this.jsonNode = this.sourceFile.getChildAt(0).getChildAt(0).expression;
    }
    static fromFileSystem(path, parserOptions) {
        const orig = readFileSync(path, 'utf-8');
        return new ParserInput(ts.parseJsonText(path, orig), orig, parserOptions);
    }
}
export class GeneratorInput {
    constructor(parseResult, inputPath, outputBase, generatorOptions) {
        this.parseResult = parseResult;
        this.inputPath = inputPath;
        this.outputBase = outputBase;
        this.generatorOptions = generatorOptions;
    }
    relativeToInput(x) {
        return join(this.inputPath, x);
    }
    relativeToOutput(x) {
        return join(this.outputBase, x);
    }
    get outputRelativePathToInput() {
        return relative(dirname(this.outputBase), this.inputPath).replace(/\\/g, '/');
    }
}
