// @perfsee/plugin-utils/dist/build-env.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuildEnv = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const env_ci_1 = tslib_1.__importDefault(require("env-ci"));
const utils_1 = require("@perfsee/utils");
const git_1 = require("./git");
const envs = (0, env_ci_1.default)();
function getPr() {
    try {
        const event = process.env.GITHUB_EVENT_PATH ? require(process.env.GITHUB_EVENT_PATH) : undefined;
        if (event?.pull_request) {
            return {
                number: parseInt(event.pull_request.number),
                baseHash: event.pull_request.base.sha,
                headHash: event.pull_request.head.sha,
            };
        }
    }
    catch {
        // Noop
    }
}
function getCiCommit() {
    try {
        const event = process.env.GITHUB_EVENT_PATH ? require(process.env.GITHUB_EVENT_PATH) : undefined;
        if (event?.pull_request) {
            return event.pull_request.head.sha;
        }
    }
    catch {
        // Noop
    }
}
function getCiBranch() {
    if (envs.isCi) {
        if ('isPr' in envs && envs.isPr) {
            if (envs.service === 'github' && process.env['GITHUB_HEAD_REF']) {
                return process.env['GITHUB_HEAD_REF'];
            }
            else {
                return 'prBranch' in envs && envs.prBranch ? envs.prBranch : envs.branch;
            }
        }
        else {
            return envs.branch;
        }
    }
}
async function getGitEnv() {
    if (envs.isCi) {
        if (envs.service === 'github' || envs.service === 'gitlab') {
            const commit = getCiCommit() || envs.commit;
            const commitMessage = await (0, git_1.getCommitMessage)(commit);
            const branch = getCiBranch() || envs.branch;
            const [namespace, name] = envs.slug.split('/');
            return {
                host: envs.service === 'github' ? utils_1.GitHost.Github : utils_1.GitHost.Gitlab,
                namespace,
                name,
                commit,
                branch,
                commitMessage,
                pr: getPr(),
            };
        }
        else if (
        // @ts-expect-error type is outdated
        envs.service === 'vercel') {
            return {
                namespace: process.env.VERCEL_GIT_REPO_OWNER,
                name: process.env.VERCEL_GIT_REPO_SLUG,
                commitMessage: process.env.VERCEL_GIT_COMMIT_MESSAGE,
                branch: process.env.VERCEL_GIT_COMMIT_REF,
                commit: process.env.VERCEL_GIT_COMMIT_SHA,
                host: process.env.VERCEL_GIT_PROVIDER === 'github'
                    ? utils_1.GitHost.Github
                    : process.env.VERCEL_GIT_PROVIDER === 'gitlab'
                        ? utils_1.GitHost.Gitlab
                        : utils_1.GitHost.Unknown,
            };
        }
        else {
            console.error(chalk_1.default.red(`[perfsee] Unsupported CI service ${envs.service}. We will try to support it in the future.`));
        }
    }
    const project = await (0, git_1.getProjectInfoFromGit)();
    const commit = await (0, git_1.getCurrentCommit)();
    if (!project || !commit) {
        throw new Error('Failed to get repository info');
    }
    const commitMessage = await (0, git_1.getCommitMessage)(commit);
    return {
        host: project.host,
        namespace: project.namespace,
        name: project.name,
        branch: project.branch,
        commit,
        commitMessage,
    };
}
let buildEnv = null;
function getBuildEnv() {
    if (!buildEnv) {
        buildEnv = {
            isCi: envs.isCi,
            pwd: 'root' in envs && envs.root ? envs.root : process.cwd(),
            platform: process.env.PERFSEE_PLATFORM_HOST ?? 'https://perfsee.com',
            upload: !process.env.PERFSEE_NO_UPLOAD && envs.isCi,
            git: getGitEnv().catch(() => null),
        };
    }
    return buildEnv;
}
exports.getBuildEnv = getBuildEnv;
