// @perfsee/plugin-utils/dist/options.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.initOptions = exports.getDefaultOptions = void 0;
const lodash_1 = require("lodash");
const build_env_1 = require("./build-env");
function getDefaultOptions() {
    return {
        artifactName: 'main',
        enableAudit: !!process.env.PERFSEE_AUDIT || (0, build_env_1.getBuildEnv)().isCi,
        token: process.env.PERFSEE_TOKEN,
        toolkit: 'webpack',
        htmlExclusive: true,
    };
}
exports.getDefaultOptions = getDefaultOptions;
function initOptions(userOptions) {
    const defaultOptions = getDefaultOptions();
    return (0, lodash_1.merge)(defaultOptions, userOptions);
}
exports.initOptions = initOptions;
