// @perfsee/plugin-utils/dist/reports.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateReports = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const find_cache_dir_1 = tslib_1.__importDefault(require("find-cache-dir"));
const lodash_1 = require("lodash");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const bundle_analyzer_1 = require("@perfsee/bundle-analyzer");
const build_env_1 = require("./build-env");
const formater_1 = require("./formater");
const viewer_1 = require("./viewer");
async function generateReports(stats, outputPath, options) {
    const { enableAudit, shouldPassAudit = (score) => score >= 80, failIfNotPass = false, processStats } = options;
    if (!enableAudit) {
        return;
    }
    if (processStats) {
        const processedStats = processStats(stats);
        if (processedStats) {
            stats = processedStats;
        }
        else {
            return;
        }
    }
    stats.rules = options.rules?.filter((rule) => typeof rule === 'string');
    stats.includeAuxiliary = options.includeAuxiliary;
    stats.htmlExclusive = options.htmlExclusive;
    stats.strictChunkRelations = options.strictChunkRelations;
    try {
        console.info('Start bundle analyzing');
        // @ts-expect-error we made it
        // eslint-disable-next-line no-console
        console.verbose = console.info;
        const platform = options.platform ?? (0, build_env_1.getBuildEnv)().platform;
        const { report, moduleTree, moduleReasons } = await bundle_analyzer_1.StatsParser.FromStats(stats, outputPath, console)
            .appendAuditsForLocal(options.rules?.filter((rule) => typeof rule === 'function') || [])
            .initAuditFetcher((path, init) => {
            return (0, node_fetch_1.default)(`${platform}${path}`, (0, lodash_1.merge)({
                headers: {
                    Authorization: `Bearer ${options.token}`,
                },
            }, init));
        }, (0, find_cache_dir_1.default)({ name: viewer_1.PACKAGE_NAME }))
            .parse();
        const score = (0, bundle_analyzer_1.calcBundleScore)(report.entryPoints);
        // directly output formatted content in CI
        if ((0, build_env_1.getBuildEnv)().isCi) {
            console.info((0, formater_1.formatAuditResult)(report));
            if (typeof score === 'number') {
                let pass = shouldPassAudit(score, report);
                if (pass instanceof Promise) {
                    pass = await pass;
                }
                if (!pass && failIfNotPass) {
                    console.error(chalk_1.default.red('[@perfsee/webpack] Bundle audit not pass.'));
                    process.exit(1);
                }
            }
            console.info('Finish bundle audit');
        }
        else {
            // start local server with report UI served
            await (0, viewer_1.saveReport)({
                branch: '',
                hash: '',
                report: report,
                content: moduleTree,
                moduleReasons,
            }, outputPath, options.reportOptions);
        }
    }
    catch (e) {
        console.error('Bundle audit error: ', e);
    }
}
exports.generateReports = generateReports;
