// @perfsee/plugin-utils/dist/viewer.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveReport = exports.renderReportViewer = exports.PACKAGE_NAME = void 0;
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const path_1 = require("path");
const find_cache_dir_1 = tslib_1.__importDefault(require("find-cache-dir"));
const open_1 = tslib_1.__importDefault(require("open"));
const bundle_analyzer_1 = require("@perfsee/bundle-analyzer");
exports.PACKAGE_NAME = '@perfsee/plugin-utils';
function renderReportViewer(data) {
    return `<!DOCTYPE html>
		<html lang="en" translate="no">
		<head>
		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<style>*,*::before,*::after{box-sizing:border-box}body,h1,h2,h3,h4,p,figure,blockquote,dl,dd{margin:0}ul[role="list"],ol[role="list"]{list-style:none}html:focus-within{scroll-behavior:smooth}body{min-height:100vh;text-rendering:optimizeSpeed;line-height:1.5}a:not([class]){text-decoration-skip-ink:auto}img,picture{max-width:100%;display:block}input,button,textarea,select{font:inherit}@media(prefers-reduced-motion:reduce){html:focus-within{scroll-behavior:auto}*,*::before,*::after{animation-duration:.01ms !important;animation-iteration-count:1 !important;transition-duration:.01ms !important;scroll-behavior:auto !important}}</style>
		<title>Perfsee bundle analyzer</title>
		<script>
			window.artifact = {
				id: 'local',
				branch: '${data.branch}',
				hash: '${data.hash}',
				createdAt: new Date().toISOString(),
				score: ${(0, bundle_analyzer_1.calcBundleScore)(data.report.entryPoints)}
			};
			window.bundleReport = ${JSON.stringify(data.report)?.replace(/</gu, '\\u003c')};
			window.bundleContent = ${JSON.stringify(data.content)?.replace(/</gu, '\\u003c')};
      window.bundleModuleReasons = ${JSON.stringify(data.moduleReasons)?.replace(/</gu, '\\u003c')};
		</script>
	</head>

	<body>
		<div id="app" style="min-height: 100vh;display: flex;flex-direction: column;"></div>
		<script>
      ${(0, fs_1.readFileSync)(require.resolve(`${exports.PACKAGE_NAME}/public/report.js`), 'utf-8')}
    </script>
	</body>
	</html>`;
}
exports.renderReportViewer = renderReportViewer;
function getDefaultReportFileName() {
    const cacheDir = (0, find_cache_dir_1.default)({ name: exports.PACKAGE_NAME });
    if (cacheDir) {
        (0, fs_1.mkdirSync)(cacheDir, { recursive: true });
        return (0, path_1.join)(cacheDir, `report-${Date.now()}.html`);
    }
    return undefined;
}
async function saveReport(data, outputPath, { fileName = getDefaultReportFileName(), openBrowser = true } = {}) {
    if (!fileName) {
        return;
    }
    const html = renderReportViewer(data);
    const htmlPath = (0, path_1.resolve)(outputPath, fileName);
    (0, fs_1.writeFileSync)(htmlPath, html, 'utf-8');
    console.info(`[perfsee] bundle report is written to ${htmlPath}`);
    if (openBrowser) {
        try {
            await (0, open_1.default)(htmlPath);
        }
        catch (err) {
            console.error(`[perfsee] failed to open default browser`, err);
        }
    }
    return htmlPath;
}
exports.saveReport = saveReport;
