// @perfsee/bundle-analyzer/dist/stats.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.dynamicModuleReasonTypes = exports.ModuleReasonTypes = exports.BundleToolkit = exports.OLD_WEBPACK_INTERNAL_PATH = exports.OLD_SOURCE_CODE_PATH = exports.WEBPACK_INTERNAL_PATH = exports.SOURCE_CODE_PATH = void 0;
exports.SOURCE_CODE_PATH = '(Source Code)';
exports.WEBPACK_INTERNAL_PATH = '(Webpack Internal)';
exports.OLD_SOURCE_CODE_PATH = '.';
exports.OLD_WEBPACK_INTERNAL_PATH = '(webpack)';
var BundleToolkit;
(function (BundleToolkit) {
    BundleToolkit["Webpack"] = "webpack";
    BundleToolkit["Esbuild"] = "esbuild";
    BundleToolkit["Rollup"] = "rollup";
    BundleToolkit["Rspack"] = "rspack";
})(BundleToolkit = exports.BundleToolkit || (exports.BundleToolkit = {}));
exports.ModuleReasonTypes = [
    'amd define',
    'amd require array',
    'amd require context',
    'amd require',
    'cjs require context',
    'cjs require',
    'context element',
    'delegated exports',
    'delegated source',
    'dll entry',
    'accepted harmony modules',
    'harmony accept',
    'harmony export expression',
    'harmony export header',
    'harmony export imported specifier',
    'harmony export specifier',
    'harmony import specifier',
    'harmony side effect evaluation',
    'harmony init',
    'import() context development',
    'import() context production',
    'import() eager',
    'import() weak',
    'import()',
    'json exports',
    'loader',
    'module.hot.accept',
    'module.hot.decline',
    'multi entry',
    'null',
    'prefetch',
    'require.context',
    'require.ensure',
    'require.ensure item',
    'require.include',
    'require.resolve',
    'single entry',
    'wasm export import',
    'wasm import',
    'new URL()',
    'cjs export require',
    'cjs self exports reference',
    'module decorator',
];
exports.dynamicModuleReasonTypes = [
    'import() context development',
    'import() context production',
    'import() eager',
    'import() weak',
    'import()',
    'prefetch',
    'require.ensure',
    'require.ensure item',
    'amd require array',
    'amd require context',
    'amd require',
    'module.hot.accept',
];
