// @toeverything/pdf-viewer/dist/bitmap.js
import { BitmapFormat } from '@toeverything/pdf-viewer-types';
export class Bitmap {
    static { this.format = BitmapFormat.BGRA; }
    constructor(runtime, ptr, format = Bitmap.format) {
        this.runtime = runtime;
        this.ptr = ptr;
        this.format = format;
    }
    get pointer() {
        return this.ptr;
    }
    close() {
        if (!this.ptr)
            return;
        this.runtime.closeBitmap(this.ptr);
        this.runtime.free(this.ptr);
        this.ptr = 0;
    }
    fill(left, top, width, height, color = 0xffffffff // 32 bit
    ) {
        return this.runtime.fillBitmap(this.ptr, left, top, width, height, color);
    }
    toUint8Array() {
        const stride = this.runtime.bitmapStride(this.ptr);
        const height = this.runtime.bitmapHeight(this.ptr);
        const bufferPtr = this.runtime.bitmapBuffer(this.ptr);
        const buffer = this.runtime.HEAPU8.subarray(bufferPtr, bufferPtr + stride * height);
        this.runtime.free(bufferPtr);
        return buffer;
    }
    [Symbol.dispose]() {
        this.close();
    }
}
