// @toeverything/pdf-viewer/dist/runtime.js
export const DefaultConfig = {
    version: 3,
    m_pIsolate: null,
    m_pPlatform: null,
    m_pUserFontPaths: null,
    m_v8EmbedderSlot: 0,
};
/**
 * PDFium runtime.
 */
export class Runtime {
    constructor(engine) {
        this.engine = engine;
        /**
         * Initialize the PDFium library and allocate global resources for it.
         */
        this.init = this.wasm.FPDF_InitLibraryWithConfig;
        this.deinit = this.wasm.FPDF_DestroyLibrary;
        /**
         * Gets last error code when a function fails.
         */
        this.lastErrorCode = (this.wasm.FPDF_GetLastError);
        this.version = this.wasm.FPDF_GetFileVersion;
        this.fileIdentifier = (this.wasm.FPDF_GetFileIdentifier);
        this.metaText = this.wasm.FPDF_GetMetaText;
        this.openDocument = this.wasm.FPDF_LoadMemDocument64;
        this.pageMode = (this.wasm.FPDFDoc_GetPageMode);
        this.pageCount = this.wasm.FPDF_GetPageCount;
        this.loadPage = this.wasm.FPDF_LoadPage;
        this.closePage = this.wasm.FPDF_ClosePage;
        this.pageLabel = this.wasm.FPDF_GetPageLabel;
        this.pageWidth = this.wasm.FPDF_GetPageWidthF;
        this.pageHeight = this.wasm.FPDF_GetPageHeightF;
        this.pageSize = this.wasm.FPDF_GetPageSizeByIndexF;
        this.pageRotation = (this.wasm.FPDFPage_GetRotation);
        this.pageTransparency = this.wasm.FPDFPage_HasTransparency;
        this.renderPageBitmap = this.wasm.FPDF_RenderPageBitmap;
        this.pointFromDeviceToPage = this.wasm.FPDF_DeviceToPage;
        this.pointFromPageToDevice = this.wasm.FPDF_PageToDevice;
        this.createBitmap = this.wasm.FPDFBitmap_Create;
        this.createBitmapWith = (this.wasm.FPDFBitmap_CreateEx);
        this.fillBitmap = this.wasm.FPDFBitmap_FillRect;
        this.closeBitmap = this.wasm.FPDFBitmap_Destroy;
        this.bitmapWidth = this.wasm.FPDFBitmap_GetWidth;
        this.bitmapHeight = this.wasm.FPDFBitmap_GetHeight;
        this.bitmapStride = this.wasm.FPDFBitmap_GetStride;
        this.bitmapBuffer = this.wasm.FPDFBitmap_GetBuffer;
        this.init(DefaultConfig);
    }
    get wasm() {
        return this.engine.wasmExports;
    }
    get HEAPU8() {
        return this.engine.HEAPU8;
    }
    malloc(size) {
        return this.wasm.malloc(size);
    }
    free(ptr) {
        this.wasm.free(ptr);
    }
    stringToUTF8(str, strPtr, maxBytesToRead) {
        this.engine.stringToUTF8(str, strPtr, maxBytesToRead);
    }
    stringToNewUTF8(str) {
        return this.engine.stringToNewUTF8(str);
    }
    UTF8ToString(ptr, maxBytesToRead) {
        return this.engine.UTF8ToString(ptr, maxBytesToRead);
    }
    UTF16ToString(ptr, maxBytesToRead) {
        return this.engine.UTF16ToString(ptr, maxBytesToRead);
    }
    getValue(ptr, type) {
        return this.engine.getValue(ptr, type);
    }
    setValue(ptr, value, type, noSafe) {
        return this.engine.setValue(ptr, value, type, noSafe);
    }
    /**
     * Copys bytes to WASM.
     */
    copyBytesTo(bytes) {
        const size = bytes.length;
        const bytesPtr = this.malloc(size);
        this.HEAPU8.set(bytes, bytesPtr);
        return bytesPtr;
    }
    closeDocument(ptr) {
        this.wasm.FPDF_CloseDocument(ptr);
        this.free(ptr);
    }
}
