// @electron-forge/maker-dmg/dist/MakerDMG.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakerDMG = void 0;
const node_path_1 = __importDefault(require("node:path"));
const maker_base_1 = require("@electron-forge/maker-base");
const fs_extra_1 = __importDefault(require("fs-extra"));
class MakerDMG extends maker_base_1.MakerBase {
    constructor() {
        super(...arguments);
        this.name = 'dmg';
        this.defaultPlatforms = ['darwin', 'mas'];
    }
    isSupportedOnCurrentPlatform() {
        return process.platform === 'darwin';
    }
    async make({ dir, makeDir, appName, packageJSON, targetArch, }) {
        const { createDMG } = require('electron-installer-dmg');
        const outPath = node_path_1.default.resolve(makeDir, `${this.config.name || appName}.dmg`);
        const forgeDefaultOutPath = node_path_1.default.resolve(makeDir, `${appName}-${packageJSON.version}-${targetArch}.dmg`);
        await this.ensureFile(outPath);
        const dmgConfig = {
            overwrite: true,
            name: appName,
            ...this.config,
            appPath: node_path_1.default.resolve(dir, `${appName}.app`),
            out: node_path_1.default.dirname(outPath),
        };
        await createDMG(dmgConfig);
        if (!this.config.name) {
            await this.ensureFile(forgeDefaultOutPath);
            await fs_extra_1.default.rename(outPath, forgeDefaultOutPath);
            return [forgeDefaultOutPath];
        }
        return [outPath];
    }
}
exports.default = MakerDMG;
exports.MakerDMG = MakerDMG;
