// @perfsee/utils/dist/git/host.js
"use strict";
/*
Copyright 2022 ByteDance and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.gitHostFromDomain = exports.CommonGitHost = exports.hostDomains = exports.GitHost = void 0;
var GitHost;
(function (GitHost) {
    GitHost["Unknown"] = "Unknown";
    GitHost["Github"] = "Github";
    GitHost["Gitlab"] = "Gitlab";
})(GitHost = exports.GitHost || (exports.GitHost = {}));
exports.hostDomains = {
    [GitHost.Github]: 'github.com',
    [GitHost.Gitlab]: 'gitlab.com',
    [GitHost.Unknown]: 'unknown-git-host.com',
};
class CommonGitHost {
    constructor(project) {
        this.path = `${exports.hostDomains[project.host] ?? ''}/${project.namespace}/${project.name}`;
    }
    repoUrl() {
        return `https://${this.path}`;
    }
    commitUrl(hash) {
        return `https://${this.path}/commits/${hash}`;
    }
    branchUrl(branch) {
        return `https://${this.path}/tree/${branch}`;
    }
    tagUrl(tag) {
        return `https://${this.path}/tree/${tag}`;
    }
    diffUrl(from, to) {
        return `https://${this.path}/compare/${from}...${to}`;
    }
    prUrl(number) {
        return `https://${this.path}/pull/${number}`;
    }
}
exports.CommonGitHost = CommonGitHost;
function gitHostFromDomain(domainOrHost) {
    const existing = GitHost[domainOrHost];
    if (existing) {
        return existing;
    }
    if (domainOrHost.includes('github.com')) {
        return GitHost.Github;
    }
    else if (domainOrHost.includes('gitlab.com')) {
        return GitHost.Gitlab;
    }
    return GitHost.Unknown;
}
exports.gitHostFromDomain = gitHostFromDomain;
